subprojects/instant-execution/src/main/kotlin/org/gradle/instantexecution/problems/InstantExecutionProblems.kt

```diff
@@ -0,0 +1,211 @@
+/*
+ * Copyright 2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.instantexecution.problems
+
+import org.gradle.instantexecution.InstantExecutionErrorException
+import org.gradle.instantexecution.InstantExecutionProblemException
+
+import kotlin.reflect.KClass
+
+
+sealed class PropertyProblem {
+
+    abstract val trace: PropertyTrace
+
+    abstract val message: StructuredMessage
+
+    abstract val exception: Throwable
+
+    companion object {
+
+        fun forWarning(trace: PropertyTrace, message: StructuredMessage, cause: Throwable? = null) =
+            Warning(trace, message, InstantExecutionProblemException(
+                exceptionMessageFor(trace, message), cause
+            ))
+
+        fun forError(trace: PropertyTrace, message: StructuredMessage, cause: Throwable) =
+            Error(trace, message, InstantExecutionErrorException(
+                exceptionMessageFor(trace, message), cause
+            ))
+
+        private
+        fun exceptionMessageFor(trace: PropertyTrace, message: StructuredMessage) =
+            propertyDescriptionFor(trace) + ": " + message
+    }
+
+    /**
+     * A problem that does not necessarily compromise the execution of the build.
```

Comment:
what would be an example, and why do we need to distinguish between these two kinds of problems at this point?

Reply:
We collect `Error`s when they occur and fail even if `failOnProblems` is `false`. This happens with exceptions thrown when serializing the task graph, except `IOException` that we throw immediately. I think I agree this is complexity we don't need and we should simply throw directly without reporting problems when facing such situations. Happy to make the change in this PR.

@bamboo wdyt?

Reply:
IIRC, that was introduced so something like a `StackOverflowError` during serialization was debuggable, i.e., would cause the build to fail only after the report was written.

I haven't looked at the full set of changes yet but it seems the report would be written in such cases with the new architecture so simplifying it looks like the right thing to do 👍